File: src/main/kotlin/jp/ac/titech/cs/se/refactorhub/tool/model/Log.kt
package jp.ac.titech.cs.se.refactorhub.tool.model

import com.fasterxml.jackson.databind.JsonNode

interface Log {
    val event: LogEvent
    val type: LogType
    val user: Int
    val time: Long
    val data: JsonNode
}

enum class LogEvent {
    Save,
    Discard,
    Update,
    PutElementKey,
    RemoveElementKey,
    VerifyElement,
    AppendElementValue,
    UpdateElementValue,
    RemoveElementValue,
    Fork,
    Edit,

    OpenElementLocation,
    ToggleEditingElement,
    SetDisplayedFile
}

enum class LogType {
    Client,
    Server
}

File: client/src/utils/log.ts
import apis, { LogEvent, LogType } from '@/apis'
import { logger } from '@/utils/logger'

export const log = async (
  event: LogEvent,
  type: LogType,
  data: object = {}
) => {
  try {
    await apis.logs.postLog({
      event,
      type,
      data,
    })
  } catch (e) {
    logger.log(e)
  }
}
